io-engine/src/bdev/ftl.rs

```diff
@@ -0,0 +1,258 @@
+//!
+//! The ftl bdev creates a CSAL-FTL disk.
+use std::{
+    collections::HashMap,
+    convert::TryFrom,
+    fmt::{Debug, Formatter},
+};
+
+use core::ffi::c_void;
+
+use async_trait::async_trait;
+use futures::channel::oneshot;
+use log::info;
+use nix::errno::Errno;
+use snafu::{OptionExt, ResultExt};
+use std::mem;
+use url::Url;
+
+use spdk_rs::ffihelper::errno_result_from_i32;
+use spdk_rs::{
+    libspdk::{
+        bdev_ftl_create_bdev, bdev_ftl_delete_bdev, ftl_bdev_info,
+        spdk_ftl_conf, spdk_ftl_get_default_conf,
+    },
+    UntypedBdev,
+};
+
+use crate::{
+    bdev::{dev::reject_unknown_parameters, util::uri, CreateDestroy, GetName},
+    bdev_api::{self, bdev_create, bdev_destroy, BdevError},
+    core::VerboseError,
+    ffihelper::{cb_arg, done_errno_cb, ErrnoResult},
+};
+
+pub struct Ftl {
+    /// the name of the ftl-bdev we created
+    name: String,
+    /// alias which can be used to open the bdev
+    alias: String,
+    /// uuid of the spdk bdev
+    uuid: Option<uuid::Uuid>,
+    // ftl's base bdev
+    bbdev_uri: String,
+    // ftl's cache bdev (must support VSS DIX mode)
```

Comment:
What is VSS ?
DIX is the data integrity stuff? So the cache bdev has to support blocksize + 8B, for example?
Any other requirements for this cache device? Does it need to be non-volatile so ex it can't be a ram device?

Reply:
I improved the docs comment on top of the file.

VSS = Variable Sector Size
DIX = Data Integrity Extension (aka Protection Information (PI))

The cache bdev needs to have an LBA format of 4KiB data size and 64B metadata size.
It also needs to be non-volatile memory if you don't want to lose data :)